### ISSUE TITLE
Feature request: Add a `parse_pattern` function which parses strings using regex patterns

### ISSUE THREAD

**Issue by abmyii** [hgrecco/pint#997]
I was searching for a library which could parse units based on a pattern given, e.g. with the pattern `{foot}'{inches}` and the input `10'11`, it would apply the pattern to the input and then convert the results to units. It's a shame that none seem to do it, and I think it would be great if pint had such a feature.

**Comment by jthielen** [comment]
For this, I would recommend using `ureg.preprocessors` and add your own custom regex preprocessor to convert a pattern like `{foot}'{inches}` to something like `{foot} ft + {inches} in` (which is then handled by `parse_expression`). Would this suit your needs?

**Comment by abmyii** [comment]
Honestly that seems like a very long way to solve quite a simple problem. It would probably take me some time to work out how to *concisely* change the string from `{foot}'{inches}` -> `{foot} ft + {inches} in`. And for anything more complex I can imagine it would be an abosolute nightmare.

**Comment by jthielen** [comment]
> Honestly that seems like a very long way to solve quite a simple problem. It would probably take me some time to work out how to _concisely_ change the string from `{foot}'{inches}` -> `{foot} ft + {inches} in`. And for anything more complex I can imagine it would be an absolute nightmare.

I'm not sure I understand how a regex substitution is a very long way to solve the problem; would you be able to clarify? For example, the following is a fairly concise way of doing this substitution:

```python
import re

from pint import UnitRegistry

def parse_ft_in_pattern(expression):
    return re.sub(
        r"(?P<foot>\d+)'(?P<inch>\d+)",
        r"\g<foot> ft + \g<inch> in",
        expression
    )

ureg = UnitRegistry(preprocessors=[parse_ft_in_pattern])
```

**Comment by abmyii** [comment]
> I'm not sure I understand how a regex substitution is a very long way to solve the problem

The difference between regex know-how and not. Thanks for the example!

May I ask what your concerns are about the "shortcut" function being implemented into `pint`? I feel like it is still a bit more concise than the above method:

```
ureg = UnitRegistry()
ureg.parse_pattern("10'11", r"(?P<foot>\d+)'(?P<inch>\d+)")
```

**Comment by jthielen** [comment]
> May I ask what your concerns are about the "shortcut" function being implemented into `pint`? I feel like it is still a bit more concise than the above method:
> 
> ```python
> ureg = UnitRegistry()
> ureg.parse_pattern("10'11", r"(?P<foot>\d+)'(?P<inch>\d+)")
> ```

The two concerns I had were:

1. This would require you to specify the pattern every time you needed to parse an expression in this fashion, rather than setting it once at the beginning (although, I guess this could be a feature depending on the exact use case!)
2. The general use case of "parsing strings using regex patterns" seems to be covered in the current API by the preprocessors option, and adding another method for such a use case felt like an unnecessary increase in complexity. However, that is definitely a good point about the proposed `parse_pattern` being more concise.

But, these were just my thoughts as a user/contributor...I'm not a maintainer. So, I think the question about how to move forward on this is best left to @hgrecco!

**Comment by abmyii** [comment]
> This would require you to specify the pattern every time you needed to parse an expression in this fashion, rather than setting it once at the beginning (although, I guess this could be a feature depending on the exact use case!)

This I was planning to address if the PR was accepted (allowing either a string pattern or a compile pattern).

> But, these were just my thoughts as a user/contributor...I'm not a maintainer. So, I think the question about how to move forward on this is best left to @hgrecco!

No problem, it's constructive and therefore very useful!

**Comment by abmyii** [comment]
https://github.com/hgrecco/pint/pull/998/commits/bbd379ad3cf62dd98357a01d8b484901a58370e7

**Comment by hgrecco** [comment]
In principle I am not against the idea but I do think we need to make better description of the use case. I think @jthielen makes a very good point: we already have an API for that .. in which cases it would be useful to have that.

I have seen the PR and is good .... except for the part you cite `parse` and not my own and little bit older https://github.com/hgrecco/stringparser/ 😄 , which (fun fact) it used to be a module from Lantz just like Pint before becoming a project on its own.

Jokes aside. I think is worth discussing. My suggestion would be: Write here the text that you would put in the docs, explaining `parse_pattern` and `preprocessors`. Documenting is a good way to test an idea.

**Comment by abmyii** [comment]
> I have seen the PR and is good .... except for the part you cite parse and not my own and little bit older hgrecco/stringparser smile , which (fun fact) it used to be a module from Lantz just like Pint before becoming a project on its own.

D'oh, I missed out on some brownie points!

> Jokes aside. I think is worth discussing. My suggestion would be: Write here the text that you would put in the docs, explaining parse_pattern and preprocessors. Documenting is a good way to test an idea.

Sure, I'll do that for `parse_pattern`. I'm not sure what you mean about documenting `preprocessors` though. Do you want me to document the alternate method suggested by @jthielen above?

**Comment by abmyii** [comment]
`parse_pattern`'s doc would be as such:

Parse a string with a given regex pattern that defines the units in the string. The measures will be extracted with the stated unit in the pattern string.

Example:

To extract the units from a height in the format `{feet}' {inches}"` we would do as such:

``` python
ureg = UnitRegistry()
ureg.parse_pattern('10\' 11"', r'(?P<foot>\d+)\' (?P<inch>\d+)"')
```

Patterns can also be compiled if many conversions are going to be made:
	
``` python
pattern = re.compile(r'(?P<foot>\d+)\' (?P<inch>\d+)"')
ureg = UnitRegistry()
ureg.parse_pattern('10\' 11"', pattern)
```

**Comment by abmyii** [comment]
And to document the additional feature in 31d4344: (in addition to above)

Multiple units can also be extracted from a string by specifying the `many=True` parameter:

``` python
ureg = UnitRegistry()
ureg.parse_pattern('10\' 10" - 10\' 11"', r'(?P<foot>\d+)\' (?P<inch>\d+)"', many=True)

> [10.833333333333334 <Unit('foot')>, 10.916666666666666 <Unit('foot')>]
```

**Comment by dalito** [comment]
What about the reverse direction, i.e. formatting the quantity according to the pattern?

I see some use cases for these types of quantities. Coordinates may also be expressed in degrees, minutes and seconds which is similar but not yet supported by pint if I am right.

I don't like the "many" feature. It seems to be against what pint normally does.

**Comment by abmyii** [comment]
> What about the reverse direction, i.e. formatting the quantity according to the pattern?

That's an interesting idea. Perhaps open a new issue?

> I don't like the "many" feature. It seems to be against what pint normally does.

How so?

**Comment by dalito** [comment]
> How so?
`pint.Quantity("1km or 100m")` and `ureg.parse_expression('1km or 100m')` fail. Both can hardly get a "many" option. Are there any methods in pint that return more than one quantity?

**Comment by abmyii** [comment]
I don't see why we should restrict progress in order to conform with "default behaviour". Also, I feel that the `parse_pattern` function targets a different "audience" of use than `Quantity` or `parse_expression`.

**Comment by abmyii** [comment]
IMO the function's purpose is to be a glorified preprocessor (as @jthielen pointed out above) which does all the parsing and construction of measures for you.

**Comment by abmyii** [comment]
@hgrecco Any update on this? I'm planning to use it in a project so would appreciate it if the decision of merge/not could be made soon. Thank you!

**Comment by hgrecco** [comment]
A few comments on some raised issues:
1. adding new functions/methods/properties: I think it should not be done just because we can, because there is a (maintainence) cost in increasing the API. The question is: does it really bring something new that is difficult to get right?
2. progress vs default behavior: I think the question @dalito is raising is about API consistency, which I 100% agree. In this context, I would **disagree** with
 - `pint.Quantity(s)` -> iterable of quantities 
 - `pint.parse_units(s)` -> iterable of units 
 - `pint.parse_expression(s)` -> iterable of quantities

because you would need to specify a `separator` (i.e. `,`, `;`, `or`)

Having said this, I am positive with the `parse_pattern` idea for the following reason: GLOBAL vs. LOCAL
- `preprocessors` are used to GLOBALLY translate COMMON (and undisputed) ways of writing numbers or units into expressions that pint can understand (e.g. `X square` ->  `X ** 2`)
- `parse_pattern` could be used to LOCALLY translate SPECIFIC ways of writing numbers or units into expressions that pint can understand (e.g. `10:20` ->  `10 hours + 20 minutes`). 

For example, I could envision parsing a table in which one of the columns has a particular structure.

However, I would go one step further, I would actually like to propose that something that was floating in your proposal: using a PEP3101 like syntax. Basically, I am proposing that we convert 
- `{unit_name}` into `(?[0-9]+\.?[0-9]+(?:[eE][-+]?[0-9]+)?+)`
Then we could actually use `{hour}:{minute}` (or `{foot}:{inch}`) as a pattern. 
(I think we should also define `_` as a capturing but ignore)

This brings something that is difficult to get right (regex are hard for a lot of people, that is why I wrote https://github.com/hgrecco/stringparser/)

Regarding the `many` clause, I am not so sure.  I have two concerns:
1. API consistency (see above)
2. return value consistency. I (usually) dislike functions that can return different types of values.

So for the time being I would suggest we leave it out and suggest to do it in the following way:
```python
>>> s = '10\' 10" or 10\' 11"'
>>> [ureg.parse_pattern(part, pattern) for part in complete.split(" or ")]
>>> [10.833333333333334 <Unit('foot')>, 10.916666666666666 <Unit('foot')>]
```

**Comment by abmyii** [comment]
> because you would need to specify a separator

Sorry, I don't quite follow here. My intention was that the regex would run a `finditer` on the string (i.e. it would find *every* match of the pattern when `many=True`), and the "or" separator I used in the example is absolutely arbitrary. 

> However, I would go one step further, I would actually like to propose that something that was floating in your proposal: using a PEP3101 like syntax. Basically, I am proposing that we convert

>    {unit_name} into (?[0-9]+\.?[0-9]+(?:[eE][-+]?[0-9]+)?+)
    Then we could actually use {hour}:{minute} (or {foot}:{inch}) as a pattern.
    (I think we should also define _ as a capturing but ignore)

> This brings something that is difficult to get right (regex are hard for a lot of people, that is why I wrote hgrecco/stringparser)

That would be great! In the PR I also mentioned the `parse` library (https://github.com/r1chardj0n3s/parse) which does this exactly, but I didn't want to add an external dependency. I guess we could just do it ourselves and it should work fine.

>  Regarding the many clause, I am not so sure. I have two concerns:
    1. API consistency (see above)
    2. return value consistency. I (usually) dislike functions that can return different types of values.

With regards to `2`; would it be sufficient if it returned a list in all cases?

> So for the time being I would suggest we leave it out and suggest to do it in the following way:

This is a perfectly reasonable solution too, and I've no issue with using it as an alternative in my use-case. My only concern is that for data with two or more different separators, it would be difficult to extract multiple values. E.g. extracting values from:

1.5kg **or** 2kg will be fine, if you do not **have** 3kg

(^ *me, when I go to the butchers*)

I think it would be cool (but not necessary) if `pint` could handle such scenarios too.

Thanks!

**Comment by hgrecco** [comment]
I understand that in your proposal the `separator` argument is not needed. My comment apply to the methods above that part. 

[stringparser](https://github.com/hgrecco/stringparser/) (and [parse](https://github.com/r1chardj0n3s/parse)) do actually much more than that as they can parse any type of integer/floating point/string/etc. My proposal is just to start with converting any `{ }` into a broad floating type capturing group.

**Comment by abmyii** [comment]
> I understand that in your proposal the separator argument is not needed. My comment apply to the methods above that part.

I see. I think I am misunderstanding this and would appreciate it if you could elaborate:

> because you would need to specify a separator (i.e. ,, ;, or)

> stringparser (and parse) do actually much more than that as they can parse any type of integer/floating point/string/etc. My proposal is just to start with converting any { } into a broad floating type capturing group.

I agree, that would be the next step up from what it is now. I'll implement that soon and update the PR. BTW the regex you provide seems to be incorrect:
``` python
re.findall('(?[0-9]+\.?[0-9]+(?:[eE][-+]?[0-9]+)?+)', '1')
```
Gave:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/re.py", line 222, in findall
    return _compile(pattern, flags).findall(string)
  File "/usr/lib/python3.6/re.py", line 301, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/usr/lib/python3.6/sre_compile.py", line 562, in compile
    p = sre_parse.parse(p, flags)
  File "/usr/lib/python3.6/sre_parse.py", line 855, in parse
    p = _parse_sub(source, pattern, flags & SRE_FLAG_VERBOSE, 0)
  File "/usr/lib/python3.6/sre_parse.py", line 416, in _parse_sub
    not nested and not items))
  File "/usr/lib/python3.6/sre_parse.py", line 752, in _parse
    len(char) + 1)
sre_constants.error: unknown extension ?[ at position 1
```

**Comment by hgrecco** [comment]
regex are hard. :-D This should do it

([+-]?[0-9]+(\.[0-9]+)?([Ee][+-]?[0-9]+)?)

**Comment by abmyii** [comment]
> regex are hard. :-D This should do it

You can say that again!

The new regex (which I also came across on SO) gives a bit of a weird output, though:
``` python3
>>> re.findall('([+-]?[0-9]+(.[0-9]+)?([Ee][+-]?[0-9]+)?)', '1.3123')
[('1.3123', '.3123', '')]
```

**Comment by keewis** [comment]
the docstring of `re.findall` says:
> If one or more capturing groups are present in the pattern, return a list of groups; this will be a list of tuples if the pattern has more than one group.

therefore the regex should be:
```python
In [4]: re.findall('([+-]?[0-9]+(?:.[0-9]+)?(?:[Ee][+-]?[0-9]+)?)', '1.3123')                                                                                                                                                   
Out[4]: ['1.3123']
```

**Comment by abmyii** [comment]
Thank you @keewis!

**Comment by abmyii** [comment]
Regex is awesome! I found and used this (https://stackoverflow.com/questions/7191209/python-re-sub-replace-with-matched-content) when implementing the aforementioned change (see PR - fed4a0a). I had to significantly overhaul my approach, however. Changes:

- Deprecate the ability to handle compiled regexes as the regex has is changed each time. There should be a creative solution if we give it enough thought.

- Disabling preprocessors - this was done since strings such as `100, 10` are interpreted as `100*10`, ect. Adding them back will make the function a lot less flexible at parsing (which, I guess, is the point).

- Returns individual units rather than results `sum`med together. This is to also increase flexibility - otherwise strings such as `100kg 10lb`, when parsed with `{kg} {lb}`, are interpreted as one item.

From:
``` python3
In [4]: ureg.parse_pattern("10kg 11lb", r"{kg}kg {lb}")                                                                                                                         
Out[4]: 14.98951607 <Unit('kilogram')>
```

To:
``` python3
In [3]: ureg.parse_pattern("10kg 11lb", r"{kg}kg {lb}lb")                                                                                                                       
Out[3]: [10.0 <Unit('kilogram')>, 11.0 <Unit('pound')>]
```

<br/>
<br/>

> (I think we should also define _ as a capturing but ignore)

@hgrecco When would this be necessary?

Also, with regards to the differing output when `many` is true/false, could we standardise it by returning nested lists in all cases? That way the output is the same and the functionality is still kept.

**Comment by hgrecco** [comment]
> Deprecate the ability to handle compiled regexes as the regex has is changed each time. There should be a creative solution if we give it enough thought.

You can use a lrucache from the pattern into the compiled regex

I will play with this and come back to you. I am still unsure about `many`

**Comment by abmyii** [comment]
> You can use a lrucache from the pattern into the compiled regex

Never heard of it before today! How would that work, and would it modify the compiled regex?

**Comment by hgrecco** [comment]
An lrucache allows you to memorize results from a function. There are multiple ways to implement this in python but the easiest way would be to define a fuction in `util`

```python
impoer functools

 @functools.lru_cache
def pattern_to_regex(pattern):
     # do your thing
     return compiled_regex
```

And then use this function in `parse_pattern`. This will speed up when things when doing something like this.

```python
for row in rows:
    out = ureg.parse_pattern(row, r"{kg}kg {lb}lb")  
    # do something
``` 

as the regex pattern will be contructed and compiled only once.

**Comment by hgrecco** [comment]
An lrucache allows you to memorize results from a function. There are multiple ways to implement this in python but the easiest way would be to define a fuction in `util`

```python
impoer functools

 @functools.lru_cache
def pattern_to_regex(pattern):
     # do your thing
     return compiled_regex
```

And then use this function in `parse_pattern`. This will speed up when things when doing something like this.

```python
for row in rows:
    out = ureg.parse_pattern(row, r"{kg}kg {lb}lb")  
    # do something
``` 

as the regex pattern will be contructed and compiled only once.

**Comment by abmyii** [comment]
> An lrucache allows you to memorize results from a function.  There are multiple ways to implement this in python but the easiest way  would be to define a fuction in util

How would this work on pre-compiled regexes that are passed to the `parse_pattern` function?

**Comment by abmyii** [comment]
@hgrecco Any updates?

**Comment by hgrecco** [comment]
I will merge this, but keeping `many` as provisional API as it is still not clear to me that is better than splitting before parsing. 

Thanks a lot for your effort int he commit and argumentation.

**Comment by abmyii** [comment]
Ah great news! I hope I wasn't too hard headed whilst arguing!

**Comment by hgrecco** [comment]
Please:
- Label del `many` argument as provisional
- squash all commits to a single one to avoid polluting the log 
- force push to your branch

**Comment by abmyii** [comment]
> Label del many argument as provisional

In the doc-string? Also, I haven't yet implemented the lru_cache for compiled regexes.

**Comment by hgrecco** [comment]
Yes, in the docstring

Regarding the `lru_cache`, if you can do it it would be great

(Just as a side note, if I remember correctly Python already has a cache that maps regex to compiled regex. So this would be actually a cache from the PEP3101 like regex to compiled regex)

**Comment by abmyii** [comment]
(Before) Without lrucache:
``` python3
In [1]: import pint                                                                                                                                                             

In [2]: import re                                                                                                                                                               

In [3]: ureg = pint.UnitRegistry()
                                                                                                                                              
In [4]: %timeit ureg.parse_pattern("1.5kg or 2kg will be fine, if you do not have 3kg", r"{kg}kg", many=True)                                                                   
597 µs ± 33.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

(After) With lrucache:

``` python3
In [1]: import pint                                                                                                                                                             

In [2]: import re                                                                                                                                                               

In [3]: ureg = pint.UnitRegistry()                                                                                                                                              

In [4]: %timeit ureg.parse_pattern("1.5kg or 2kg will be fine, if you do not have 3kg", re.compile(r"{kg}kg"), many=True)                                                       
607 µs ± 6.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit ureg.parse_pattern("1.5kg or 2kg will be fine, if you do not have 3kg", r"{kg}kg", many=True)                                                                   
604 µs ± 8.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by hgrecco** [comment]
Th improvement is inexistent but I still think is a good idea to split the function as you did as it is easier to read.

I will merge this as soon as you add this to the documentation. I think that within the "String Parsing" section of the tutorila will be great.

**Comment by abmyii** [comment]
Done - https://github.com/hgrecco/pint/pull/998/commits/9b673210526d9a21fd51c2f8d8a401767948e047. It seems that the GitHub RST editor stripped some of the trailing characters/whitespace, thus the edits to other parts of the document.

Should the `lru_cache` be left at `1024`?

If all is good then I'll squash the commits and force-push the branch.

**Comment by hgrecco** [comment]
Just use 
```python
@lru_cache
def ...
```

that will use the default value.

**Comment by abmyii** [comment]
Squashed and force-pushed: https://github.com/hgrecco/pint/commit/95efc46795c2ec22d78c41cd81394728c0488cd1

Thank you all (@hgrecco especially) for the support throughout!
